docs/input/docs/usage/command-line.md

```diff
@@ -119,14 +119,65 @@ Will iterate through each file and update known attributes (`AssemblyVersion`,
 
 `/overrideconfig [key=value]` will override appropriate key from 'GitVersion.yml'.
 
-At the moment only `tag-prefix` option is supported. Read more about [Configuration](../configuration).
+Multiple options are separated by semicolon: `key1=value1;key2=value2`.
+
+To have **space characters** as a part of `value`, `value` has be enclosed with double quotes - `key="My value"`.
+
+Double quote character inside of the double quoted `value` has to be be escaped with a backslash '\\' - `key="My \"escaped-quotes\""`.
+
+If closing double-qoute character is missing, it will be implicitily added at the very end of the command line argument - `key="My Value;key2=2(")`.
```

Comment:
Let's just throw an exception when parsing fails instead. Autocorrection just leads to more errors and confusion, imho.

Reply:
For this behavior I also consulted MSBuild implementation. Specifically `[QuotingUtilities.cs](https://github.com/dotnet/msbuild/blob/e3eac4c45510dcb85a4a2aea323dae81189f60bb/src/Shared/QuotingUtilities.cs)` class:

```
        /*
         * Quoting Rules:
         * 
         * A string is considered quoted if it is enclosed in double-quotes. A double-quote can be escaped with a backslash, or it
         * is automatically escaped if it is the last character in an explicitly terminated quoted string. A backslash itself can
         * be escaped with another backslash IFF it precedes a double-quote, otherwise it is interpreted literally.
         * 
         * e.g.
         *      abc"cde"xyz         --> "cde" is quoted
         *      abc"xyz             --> "xyz" is quoted (the terminal double-quote is assumed)
         *      abc"xyz"            --> "xyz" is quoted (the terminal double-quote is explicit)
         * 
         *      abc\"cde"xyz        --> "xyz" is quoted (the terminal double-quote is assumed)
         *      abc\\"cde"xyz       --> "cde" is quoted
         *      abc\\\"cde"xyz      --> "xyz" is quoted (the terminal double-quote is assumed)
         * 
         *      abc"""xyz           --> """ is quoted
         *      abc""""xyz          --> """ and "xyz" are quoted (the terminal double-quote is assumed)
         *      abc"""""xyz         --> """ is quoted
         *      abc""""""xyz        --> """ and """ are quoted
         *      abc"cde""xyz        --> "cde"" is quoted
         *      abc"xyz""           --> "xyz"" is quoted (the terminal double-quote is explicit)
         * 
         *      abc""xyz            --> nothing is quoted
         *      abc""cde""xyz       --> nothing is quoted
         */
```

I assumed this is expected behavior when parsing command line quotes.

Reply:
It may be expected when working with MSBuild, but GitVersion is not an MSBuild utility (although it has an MSBuild abstraction), but its own self-sustained tool. It's more important that GitVersion provides a good, clean interface familiar to developers used to a wide variety of CLI tools and not just Microsoft developers only familiar with MSBuild.

Reply:
I think MSBuild did try not invent new way of parsing, just modeled their implementation on most common scenario, but this is no problem. We can easily throw the exception.

Reply:
Exception please 👍🏼

Reply:
While implementing throwing of exception, I discovered why MSBuild parsing of quotes was implemented with implicit termination at the end of string. It has same behavior as .NET Command Line applications. I suspect this to be convention and that other command-line apps have similar behavior (eg. C++ command line apps).

If I pass following, real-life parameters to the GitVersion.exe with double-quote missing at the very end:

`/overrideconfig assembly-versioning-format="{Major}.{Minor}.{Patch}.{env:BUILD_NUMBER ?? 0}`

`Main(string[] args)` method receives `args` where open double-quote is implicitly terminated and removed. More importantly, last parameter has spaces ` ?? 0`, but was not split, proving that the double-quote was indeed assumed at the very end of the line. Screenshot will make it more clear:

![image](https://user-images.githubusercontent.com/4405210/106053637-51bf3a80-60eb-11eb-9cea-1a066711ce60.png)

So, once I have implemented throwing of the exception, what happens is that unit-tests are throwing, but same parameters missing closing double-quotes in the command line just work normally. Therefore we just introduced inconsistency.

Additionally, now that I looked up the source code again after a few weeks - reason why I looked at MSBuild command-line parsing logic in the first place was that you have MSBuildTask implementation of `GitVersion`. This one can be called via `msbuild`, something like:

`msbuild /target:GitVersion /property:key=value;key=value`

So, my thinking was that more similarities in format would help users.

Reply:
If it's difficult keeping the behavior consistent, then don't throw an exception. But I'm adamant we should not use `;` as a separator if we can avoid it.

Reply:
> If it's difficult keeping the behavior consistent, then don't throw an exception. But I'm adamant we should not use `;` as a separator if we can avoid it.

In any case we need to have a solution that will be easy to migrate in v6. There we use the System.CommandLine

Reply:
Guys,

Requirements you put forward do not add up for me, from what I see looking at the current code.

> But I'm adamant we should not use ; as a separator if we can avoid it.

You can use other special character, instead of `;` for the sake of being different from MSBuild, but I ';' is probably the cleanest of all the special separator characters.

In order to use space character complete parsing logic of `ArgumentParser` would have to be rewritten.

Current `ArgumentParser` implementation has a lot of explicit branching decisions per switch value. Each switch value has it's own parsing logic and there is very little general-purpose parsing code. Check `ParseArguments` method. Further, there is even explicit check `EnsureArgumentValueCount` to ensure each switch accepts single argument, which is in contradiction with space character being used to separate child-options for a switch. We would screw up existing validation logic.

Regarding portability of this pull request to v6 command line parsing:

> In any case we need to have a solution that will be easy to migrate in v6. There we use the [System.CommandLine](https://github.com/dotnet/command-line-api)

All that would have to be done to make current `ParseOverrideConfig(Arguments arguments, string value)` method work in version 6 would be to make it accept array `string[] values` instead of value, and remove line inside of method that splits `value` based on `;` character.

> If it's too involved, let's postpone this to GitVersion 6 where we will redo the entire CLI (see #2275).

I'm afraid that postponing this change to GitVersion 6 would mean throwing current implementation away (code + unit-tests) and start everything from scratch.  This pull request started from idea to add single additional `/overrideconfig` option, extended to add 26 extra options, which based on my free time, I could do. For re-implementing all command line parsing in v5 or re-implementing pull request in v6, I cannot commit myself.

My proposal would be to integrate this as is. It will be breaking change to switch from `key1=value1;key2=value2` to space characters `key1=value1 key2=value2` in v6, but you will have breaking changes in any case, because complete parsing logic will be different.

Reply:
Sorry for the late reply. I wonder, is it also impossible to repeat the `/overrideconfig` argument?

Reply:
I commented on the possibility of repeated options here:
https://github.com/GitTools/GitVersion/pull/2524#discussion_r565340553

The way parsing is implemented at the moment, second `/output` is not taken into consideration:

```
gitversion /output json /output buildserver
```

As far as I investigated the code, allowing multiple instances of the same option would also be global parser change.

Reply:
`-output json -output buildserver` was implemented in #2128 and must be working as the following test (and many more) is successful?

https://github.com/GitTools/GitVersion/blob/1fd097957a0160ab0558867d36fe89917468b17b/src/GitVersion.App.Tests/ArgumentParserTests.cs#L173-L180

Reply:
@asbjornu This was very helpful example. After investigation implementation of this example, I managed to implement specifying multiple `overrideconfig` options using repeated `/overrideconfig` arguments, with only a few changes.